client/core: order status sync after preimage miss

chappjc opened the pull request:
When reconciling an order that was in epoch status when the connection was lost and not restored before the preimage request (a preimage miss), be sure to unlock the funding coins when reconciling the status of the order to revoked.

There are two variants of this failure mode:
- Client reconnects after that and the order status is already **revoked**.  The first commit that adds a missing `trade.maybeReturnCoins` to the `updateOrder` closure addresses this.
- Client reconnects while the preimage collection is still active and the order status is still **epoch**, but they've still missed their preimage request (server doesn't re-send it).  This is addressed by flagging epoch orders in handleReconnect for a status check after a delay.

This PR also adds some additional edge case handling to `reconcileTrades`.  One fix handles a malicious or malfunctioning server that returns different orders than requested in the order status result.  The other deals with orders that get no order status result from the server.

chappjc commented on client/asset/btc/btc.go:
```diff
@@ -2047,7 +2047,7 @@ func fund(utxos []*compositeUTXO, enough func(uint64, uint64) bool) (
 		if !tryUTXOs(0) {
 			return 0, 0, nil, nil, nil, nil, fmt.Errorf("not enough to cover requested funds. "+
-				"%s BTC available in %d UTXOs", amount(sum), len(coins))
+				"%s available in %d UTXOs", amount(sum), len(coins))
```
This method is used by all the clones and it's not aware of the asset symbol.  It could be passed down, but it gets kind of ridiculous (tried it), and ultimately the top level caller knows the asset.

Related, which needs to be addressed because of how these errors are wrapped by the callers: https://github.com/decred/dcrdex/issues/1488#issuecomment-1153432115

chappjc commented on client/core/core.go:
```diff
@@ -843,7 +846,9 @@ func (dc *dexConnection) reconcileTrades(srvOrderStatuses []*msgjson.OrderStatus
-		// canceled, or revoked, return the remaining quantity.
+		// canceled, or revoked, and there are no active matches, return the
+		// locked funding coins and any refund/redeem reserves.
+		trade.maybeReturnCoins()
```
A very longest standing bug in dex.  An order being reconciled from epoch -> revoked/canceled status needs it's coins returned.

chappjc commented on client/core/core.go:
```diff
@@ -6718,6 +6764,10 @@ func handleMatchProofMsg(c *Core, dc *dexConnection, msg *msgjson.Message) error
 
+	// May need to reconcile epoch status orders from old epochs (missed
+	// preimage request and reconnected while order still in epoch status).
+	go checkMarketEpochOrders(dc, note.MarketID, note.Epoch)
```
This bug is stranger.  You can end up with an order that stays in **epoch status** forever.

This is when you submit an order, lose connection, server sends a preimage request while you are gone, you come back before the end of preimage collection for that epoch (order is still epoch status according to server during reconciliation), but since it is a preimage miss, the order gets neither a `nomatch` or a `match` request from the server where the order status normally advances.  Unless you reconnect or restart to force another order status reconciliation, there is nothing to clean up these orphaned epoch status orders.

It's hard to test this, but you can manage the timing if you retry several times, usually with a CTRL+Z to suspend, and then wait a bit and resume (fd), hopefully after the request has failed but the server hasn't yet changed the order status to revoked yet.

The solution here is when new epochs are reported in the match_proof, we check all our epoch status trades to see if they are from a bygone epoch.

chappjc replied:
No because it's exactly the problem that the reconnect happens before the order changes from epoch status to revoked server side.

chappjc replied:
This issue of relying on the match_proof is troublesome.  There could be regular checks in tick, but the intervals would be larger and the check would be across all markets.  Alternatively `handleReconnect` could *schedule* a check for a couple epochs in the future, just to be safe.

buck54321 replied:
> Alternatively `handleReconnect` could _schedule_ a check for a couple epochs in the future, just to be safe.

That's kinda where my head was going too. On reconnect, we can mark all epoch orders for a check upon some conditions set of conditions in tick.

chappjc replied:
I'll take a crack at that.  On reconnect, if there are any epoch status orders (where we might have missed the preimage request), schedule some `checkMarketEpochOrders` for the relevant markets...

chappjc commented on client/core/core.go:
```diff
@@ -932,10 +937,35 @@ func (dc *dexConnection) reconcileTrades(srvOrderStatuses []*msgjson.OrderStatus
 			var oid order.OrderID
 			copy(oid[:], srvOrderStatus.ID)
 			trade := knownActiveTrades[oid] // no need to lock dc.tradeMtx
+			if trade == nil {
+				dc.log.Warnf("Server reported status for order %v that we did not request.", oid)
```
This is a defense against a hostile or buggy server, which could cause client to panic.

chappjc commented on client/core/core.go:
```diff
@@ -932,10 +937,35 @@ func (dc *dexConnection) reconcileTrades(srvOrderStatuses []*msgjson.OrderStatus
+				}
+			}
+			// No result for this order.
+			dc.log.Warnf("Server did not report status for order %v", req.OrderID)
```
I believe we've encountered cases where non-existent (server-side) orders would be handled by this.  I'm not sure how a client would have an order but not server, but I think this has happened and it should be handled.

JoeGruffins reviewed (approved):
Changes make sense and testing well.

chappjc commented on client/core/core.go:
```diff
@@ -6710,6 +6740,22 @@ func (c *Core) handleConnectEvent(dc *dexConnection, status comms.ConnectionStat
+		status := trade.metaData.Status
+		trade.mtx.RUnlock()
+		if status == order.OrderStatusEpoch && trade.epochIdx() < epochIdx { // not == because match msgs are after match_proof and epoch_report
+			dc.reconcileTrades(nil)
```
That's correct.  It's the reason for the `previousStatus == newStatus` check at your other comment. https://github.com/decred/dcrdex/pull/1849/files/b7a6c0fa6735624c08caf32ea20181282a081677#r975170938
Is that what you were saying we would never reach?

buck54321 replied:
Ah. It's all coming together. We don't necessarily want to log those.

chappjc replied:
Ditched this in favor of jacking into handleReconnect and recording epoch orders to check after a delay.

chappjc commented on client/core/core.go:
```diff
@@ -774,6 +774,124 @@ func (dc *dexConnection) compareServerMatches(srvMatches map[order.OrderID]*serv
+// updateOrderStatus updates the order's status, cleaning up any associated
+// cancel orders, unlocking funding coins and refund/redemption reserves, and
+// updating the order in the DB. The trackedTrade's mutex must be write locked.
+func (dc *dexConnection) updateOrderStatus(trade *trackedTrade, newStatus order.OrderStatus) {
```
These two new methods are essentially just factored out of `reconcileTrades`

chappjc commented on client/core/core.go:
```diff
@@ -6616,6 +6670,33 @@ func (c *Core) handleReconnect(host string) {
+	}
+	for mkt := range epochOrders {
+		trades := epochOrders[mkt] // don't capture loop var below
+		time.AfterFunc(
```
`tick`s are a little too infrequent for my taste.

This should be fine even if there are frequent reconnects because the `func` is mostly a no-op if the orders are no longer epoch status.

It's also appealing to grab a local slice of the orders rather than having to add more fields to `dexConnection` and then iterate through `dc.trades` in `tick`.  I'm open to exploring a specific suggestion for an approach that uses `tick` however.

buck54321 commented on client/core/core.go:
```diff
@@ -6616,6 +6670,33 @@ func (c *Core) handleReconnect(host string) {
+	for mkt := range epochOrders {
+		trades := epochOrders[mkt] // don't capture loop var below
+		time.AfterFunc(
+			preimageReqTimeout+time.Duration(dc.marketEpochDuration(mkt))*time.Millisecond,
```
Should check the `Core.ctx.Err()` before doing anything here.

chappjc commented:
#### With just the client-side handling (without the second server commit)

**dexc**

```
2022-09-23 12:32:48.355 [INF] CORE[btc]: Funding 0.0135 btc order with coins [01ed7f5d47ed932ad09aa437b92c6f23864b2556baa25f66a242340260de183d:1] worth 1.1
2022-09-23 12:32:48.373 [DBG] CORE: notify: |POKE| (order) Order placed - Buying 10.00000000 DCR, rate = 0.00135 (488a45ad) - Order: 488a45ad4373309ff601ab5407fd1f67242f21e54f72341e681501585267bd9b
^Z
[1]+  Stopped                 ./dexc --simnet
<wait for server to send preimage request and for the websocket conn to drop, but before revoke>
 $  fg
./dexc --simnet
2022-09-23 12:33:18.515 [DBG] CORE: notify: |POKE| (conn) Server connected - 127.0.0.1:17273 is connected
2022-09-23 12:33:18.515 [INF] CORE[wss://127.0.0.1:17273/ws]: Successfully reconnected.
2022-09-23 12:33:18.516 [DBG] CORE: Server 127.0.0.1:17273 broadcast timeout 1m0s. Tick interval 7.5s
2022-09-23 12:33:18.523 [DBG] CORE: Authenticated connection to 127.0.0.1:17273, acct bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77, 1 active orders, 0 active matches, score 4 (suspended = false)
2022-09-23 12:33:18.526 [TRC] CORE: Status reconciliation not required for order 488a45ad4373309ff601ab5407fd1f67242f21e54f72341e681501585267bd9b, status "epoch", server-reported status "epoch"
<NOTE still epoch status because matching not done yet, but missed the preimage request>
...
<AfterFunc fires to reconcile now that the epoch is passed and preimage collection is done>
2022-09-23 12:33:53.527 [DBG] CORE: Requesting statuses for 1 orders from DEX 127.0.0.1:17273
2022-09-23 12:33:53.528 [DBG] CORE[btc]: returning coins [01ed7f5d47ed932ad09aa437b92c6f23864b2556baa25f66a242340260de183d:1]
2022-09-23 12:33:53.530 [WRN] CORE: Order 488a45ad4373309ff601ab5407fd1f67242f21e54f72341e681501585267bd9b updated from recorded status "epoch" to new status "revoked" reported by DEX 127.0.0.1:17273
2022-09-23 12:33:53.531 [WRN] CORE: notify: |WARNING| (order) Order status update - Status of order 488a45ad revised from epoch to revoked - Order: 488a45ad4373309ff601ab5407fd1f67242f21e54f72341e681501585267bd9b
2022-09-23 12:33:56.853 [INF] CORE: Retiring inactive order 488a45ad4373309ff601ab5407fd1f67242f21e54f72341e681501585267bd9b in status revoked
```

**dcrdex (without added revoke_order note)**

```
2022-09-23 12:32:48.367 [TRC] MKT: Received order 488a45ad4373309ff601ab5407fd1f67242f21e54f72341e681501585267bd9b at 2022-09-23 17:32:48.367 +0000 UTC
2022-09-23 12:33:00.001 [TRC] MKT: Preimage request sent for order 488a45ad4373309ff601ab5407fd1f67242f21e54f72341e681501585267bd9b
2022-09-23 12:33:17.087 [ERR] COMM[WS]: Websocket receive error from peer 127.0.0.1: read tcp4 127.0.0.1:17273->127.0.0.1:48218: i/o timeout (*websocket.netError)
2022-09-23 12:33:17.087 [TRC] COMM[WS]: Sent 7 and dropped 0 messages to 127.0.0.1 before shutdown.
2022-09-23 12:33:17.087 [DBG] AUTH: Link down: id=0, ip=127.0.0.1.
2022-09-23 12:33:17.087 [TRC] COMM: Disconnected websocket client 127.0.0.1
2022-09-23 12:33:18.515 [DBG] COMM: Starting websocket handler for 127.0.0.1:36290
2022-09-23 12:33:18.515 [TRC] COMM: New websocket client 127.0.0.1
2022-09-23 12:33:18.515 [TRC] COMM[WS]: Starting websocket messaging with peer 127.0.0.1
2022-09-23 12:33:18.519 [TRC] AUTH: User bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77 cancellation rate is now 0.00% (0 cancels : 2 successes). Violation = false
2022-09-23 12:33:18.519 [DBG] AUTH: User bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77 score = 4: 0 (violations) + 4 (2 preimage misses) - 0 (0 successes)
2022-09-23 12:33:18.520 [INF] AUTH: Authenticated account bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77 from 127.0.0.1 with 1 active orders, 0 active matches (score = 4, suspended = false)
2022-09-23 12:33:18.520 [DBG] AUTH: Stopped unbook timer for user bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77
2022-09-23 12:33:20.001 [INF] MKT: Collected 0 valid order preimages, missed 1. Commit checksum: 65eac13dce7f69906fef9d73b78b303db728d1461671f0cb7c35f8cdaffa763a
2022-09-23 12:33:20.001 [INF] MKT: No preimage received for order 488a45ad4373309ff601ab5407fd1f67242f21e54f72341e681501585267bd9b from user bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77. Recording violation and revoking order.
2022-09-23 12:33:20.003 [DBG] AUTH: Recorded order 8880ab65d6b47850c20fcca35cde522a0f011cdda814263978557f75f00e5e74 that has finished processing: user=bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77, time=1663954400002, target=488a45ad4373309ff601ab5407fd1f67242f21e54f72341e681501585267bd9b
2022-09-23 12:33:20.003 [TRC] AUTH: User bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77 cancellation rate is now 0.00% (0 cancels : 3 successes). Violation = false
2022-09-23 12:33:20.003 [DBG] AUTH: Registering outcome "preimage miss" (badness 2) for user bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77, new score = 6
2022-09-23 12:33:53.528 [TRC] AUTH: 1 results for 1 requested order statuses, acct = bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77
```

#### With the new server's revoke order note added in the second commit:

**dexc**

```
2022-09-23 12:01:30.478 [INF] CORE[btc]: Funding 0.027 btc order with coins [01ed7f5d47ed932ad09aa437b92c6f23864b2556baa25f66a242340260de183d:1] worth 1.1
2022-09-23 12:01:30.490 [DBG] CORE: notify: |POKE| (order) Order placed - Buying 20.00000000 DCR, rate = 0.00135 (5d4dcf9d) - Order: 5d4dcf9ded46d8f06cbad7c3a244037411903f8ada4f2337c2196885095a82c2
^Z
[1]+  Stopped                 ./dexc --simnet
<same wait>
$  fg
./dexc --simnet
2022-09-23 12:01:47.013 [INF] CORE[wss://127.0.0.1:17273/ws]: Attempting to reconnect to wss://127.0.0.1:17273/ws...
2022-09-23 12:01:47.013 [DBG] CORE: Received preimage request for order 5d4dcf9ded46d8f06cbad7c3a244037411903f8ada4f2337c2196885095a82c2 with known commitment cb85699c7b2b4519f4f75af47eca562797c9bb3121916aab620a6d58efd3a2b5
2022-09-23 12:01:47.013 [ERR] CORE: async processPreimageRequest for 5d4dcf9ded46d8f06cbad7c3a244037411903f8ada4f2337c2196885095a82c2 failed: preimage send error: cannot send on a broken connection
2022-09-23 12:01:47.031 [DBG] CORE: notify: |POKE| (conn) Server connected - 127.0.0.1:17273 is connected
2022-09-23 12:01:47.031 [INF] CORE[wss://127.0.0.1:17273/ws]: Successfully reconnected.
2022-09-23 12:01:47.032 [DBG] CORE: Server 127.0.0.1:17273 broadcast timeout 1m0s. Tick interval 7.5s
2022-09-23 12:01:47.039 [DBG] CORE: Authenticated connection to 127.0.0.1:17273, acct bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77, 1 active orders, 0 active matches, score 2 (suspended = false)
2022-09-23 12:01:47.042 [TRC] CORE: Status reconciliation not required for order 5d4dcf9ded46d8f06cbad7c3a244037411903f8ada4f2337c2196885095a82c2, status "epoch", server-reported status "epoch"
<NOTE still epoch status because matching not done yet, but missed the preimage request>
...
<then server sends a revoke_order, which we receive, so no waiting to reconcile order status>
2022-09-23 12:02:05.005 [WRN] CORE: Revoking order 5d4dcf9ded46d8f06cbad7c3a244037411903f8ada4f2337c2196885095a82c2
2022-09-23 12:02:05.006 [DBG] CORE[btc]: returning coins [01ed7f5d47ed932ad09aa437b92c6f23864b2556baa25f66a242340260de183d:1]
2022-09-23 12:02:05.007 [ERR] CORE: notify: |ERROR| (order) Order revoked - Order 5d4dcf9d on market dcr_btc at 127.0.0.1:17273 has been revoked by the server - Order: 5d4dcf9ded46d8f06cbad7c3a244037411903f8ada4f2337c2196885095a82c2
2022-09-23 12:02:05.009 [TRC] CORE[btc][SPV]: Bals: spendable = 6.6 BTC (6.6 BTC trusted, 0 BTC untrusted, 0 BTC assumed locked), immature = 0 BTC
2022-09-23 12:02:05.010 [TRC] CORE: notify: |DATA| (balance) balance updated
2022-09-23 12:02:05.010 [TRC] CORE[dcr_btc][book]: Validating match proof note for epoch 110930166 (dcr_btc) with 0 preimages and 1 misses.
2022-09-23 12:02:11.853 [INF] CORE: Retiring inactive order 5d4dcf9ded46d8f06cbad7c3a244037411903f8ada4f2337c2196885095a82c2 in status revoked
```

**dcrdex (with revoke_order note added)**

```
2022-09-23 12:01:30.484 [TRC] MKT: Received order 5d4dcf9ded46d8f06cbad7c3a244037411903f8ada4f2337c2196885095a82c2 at 2022-09-23 17:01:30.484 +0000 UTC               
                                                                                                                                                                      
2022-09-23 12:01:45.001 [TRC] MKT: Preimage request sent for order 5d4dcf9ded46d8f06cbad7c3a244037411903f8ada4f2337c2196885095a82c2                                   
2022-09-23 12:01:45.351 [ERR] COMM[WS]: Websocket receive error from peer 127.0.0.1: read tcp4 127.0.0.1:17273->127.0.0.1:39902: i/o timeout (*websocket.netError)    
2022-09-23 12:01:45.351 [DBG] AUTH: Link down: id=4, ip=127.0.0.1.                                                                                                    
2022-09-23 12:01:45.352 [TRC] COMM[WS]: Sent 15 and dropped 0 messages to 127.0.0.1 before shutdown.                                                                  2022-09-23 12:01:45.352 [TRC] COMM: Disconnected websocket client 127.0.0.1                                                                                           
                                                                                                                                                                      2022-09-23 12:01:47.030 [DBG] COMM: Starting websocket handler for 127.0.0.1:43134                                                                                    
2022-09-23 12:01:47.030 [TRC] COMM: New websocket client 127.0.0.1                                                                                                    2022-09-23 12:01:47.030 [TRC] COMM[WS]: Starting websocket messaging with peer 127.0.0.1                                                                              
2022-09-23 12:01:47.034 [TRC] AUTH: User bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77 cancellation rate is now 0.00% (0 cancels : 1 successes). Violation = false                                                                                                                                                       
2022-09-23 12:01:47.034 [DBG] AUTH: User bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77 score = 2: 0 (violations) + 2 (1 preimage misses) - 0 (0 successes)                                                                            
2022-09-23 12:01:47.035 [INF] AUTH: Authenticated account bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77 from 127.0.0.1 with 1 active orders, 0 active matches (score = 2, suspended = false)                                                                                                                             
2022-09-23 12:01:47.035 [DBG] AUTH: Stopped unbook timer for user bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77                                    2022-09-23 12:02:05.002 [INF] MKT: Collected 0 valid order preimages, missed 1. Commit checksum: 9e271484ab22be1cec2a82c5e81261974a8b2dceccd56eb71273d916c04691b8     
2022-09-23 12:02:05.002 [INF] MKT: No preimage received for order 5d4dcf9ded46d8f06cbad7c3a244037411903f8ada4f2337c2196885095a82c2 from user bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77. Recording violation and revoking order.                                                                                      
2022-09-23 12:02:05.004 [DBG] AUTH: Recorded order 5e467993e946a501eed07b96a9bfcd88279a5285f54a1c6c14874efa93909bc6 that has finished processing: user=bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77, time=1663952525003, target=5d4dcf9ded46d8f06cbad7c3a244037411903f8ada4f2337c2196885095a82c2
2022-09-23 12:02:05.004 [TRC] AUTH: User bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77 cancellation rate is now 0.00% (0 cancels : 2 successes). Violation = false                                                                    
2022-09-23 12:02:05.004 [DBG] AUTH: Registering outcome "preimage miss" (badness 2) for user bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77, new score = 4                                                                                                                                                                2022-09-23 12:02:05.004 [INF] MKT: Sending a 'revoke_order' notification to bd122f9af6c77af7268121b5e51f081a47bfca47f4fd14162184692e7635bb77 for order 5d4dcf9ded46d8f
06cbad7c3a244037411903f8ada4f2337c2196885095a82c2
```
